# Ejercicio 4

Usaremos el script eval.py con los modelos ya entrenados guardados en la carpeta trainedModels:

In [1]:
%run tagging/scripts/eval.py --help

Evaulate a tagger.

Usage:
  eval.py [options] -c <path> -i <file>
  eval.py -h | --help

Options:
  -c <path>     Ancora corpus path.
  -i <file>     Tagging model file.
  -p            Show progress bar.
  -m            Show confusion matrix.
  -g            Show confusion matrix with github markdown.

  -h --help     Show this screen.


In [8]:
!ls trainedModels

MNBclass SVMclass logR


## Logistic regression

In [9]:
%run tagging/scripts/eval.py -c ancora-3.0.1es -i trainedModels/logR -m

/Users/rodri/anaconda/anaconda3/lib/python3.6/site-packages/sklearn/base.py:311: UserWarning: Trying to unpickle estimator DictVectorizer from version 0.21.2 when using version 0.19.1. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)
/Users/rodri/anaconda/anaconda3/lib/python3.6/site-packages/sklearn/base.py:311: UserWarning: Trying to unpickle estimator LogisticRegression from version 0.21.2 when using version 0.19.1. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)
/Users/rodri/anaconda/anaconda3/lib/python3.6/site-packages/sklearn/base.py:311: UserWarning: Trying to unpickle estimator Pipeline from version 0.21.2 when using version 0.19.1. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)


Evaluation time 3.307827949523926

Accuracy: 91.69% / 95.01% / 61.68% (total / known / unk)

g \ m	sp000	nc0s000	da0000	aq0000	fc	nc0p000	rg	np00000	fp	cc
sp000	14.26	0.01	-	0.05	-	-	0.01	0.00	-	-	
nc0s000	0.00	11.89	0.01	0.52	-	0.02	0.02	0.09	-	-	
da0000	-	0.11	9.50	0.01	-	0.01	-	0.01	-	-	
aq0000	0.00	0.51	-	6.43	-	0.17	0.01	0.06	-	-	
fc	-	-	-	-	5.85	-	-	-	-	-	
nc0p000	-	0.35	-	0.46	-	4.61	-	0.04	-	-	
rg	0.02	0.04	0.00	0.39	-	0.00	3.11	0.03	-	0.02	
np00000	-	0.22	-	0.08	-	0.01	-	3.24	-	0.00	
fp	-	-	-	-	-	-	-	-	3.55	-	
cc	0.00	0.00	-	0.01	-	0.00	0.05	0.00	-	3.34	


## Naive bayes

In [11]:
%run tagging/scripts/eval.py -c ancora-3.0.1es -i trainedModels/MNBclass -m

/Users/rodri/anaconda/anaconda3/lib/python3.6/site-packages/sklearn/base.py:311: UserWarning: Trying to unpickle estimator DictVectorizer from version 0.21.2 when using version 0.19.1. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)
/Users/rodri/anaconda/anaconda3/lib/python3.6/site-packages/sklearn/base.py:311: UserWarning: Trying to unpickle estimator MultinomialNB from version 0.21.2 when using version 0.19.1. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)
/Users/rodri/anaconda/anaconda3/lib/python3.6/site-packages/sklearn/base.py:311: UserWarning: Trying to unpickle estimator Pipeline from version 0.21.2 when using version 0.19.1. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)


Evaluation time 182.94684600830078

Accuracy: 84.28% / 88.07% / 49.99% (total / known / unk)

g \ m	sp000	nc0s000	da0000	aq0000	fc	nc0p000	rg	np00000	fp	cc
sp000	14.31	-	0.01	0.00	0.00	0.00	0.00	0.00	-	-	
nc0s000	0.05	12.05	0.07	0.09	0.02	0.01	0.00	0.29	-	0.00	
da0000	0.00	0.14	9.52	0.00	0.00	0.00	-	0.01	-	-	
aq0000	0.22	0.88	0.19	5.11	0.14	0.23	0.01	0.40	0.00	0.01	
fc	0.00	-	-	-	5.85	-	-	0.00	-	-	
nc0p000	0.06	0.28	0.17	0.07	0.02	4.64	0.00	0.26	-	0.00	
rg	0.20	0.17	0.08	0.14	0.04	0.01	2.77	0.23	-	0.02	
np00000	0.03	0.27	0.05	0.00	0.00	0.01	0.00	3.21	-	-	
fp	0.00	-	-	-	-	-	-	-	3.55	-	
cc	0.01	0.00	0.00	-	-	0.00	0.05	0.05	-	3.29	


## Vector machine

In [1]:
%run tagging/scripts/eval.py -c ancora-3.0.1es -i trainedModels/SVMclass -m

/Users/rodri/anaconda/anaconda3/lib/python3.6/site-packages/sklearn/base.py:311: UserWarning: Trying to unpickle estimator DictVectorizer from version 0.21.2 when using version 0.19.1. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)
/Users/rodri/anaconda/anaconda3/lib/python3.6/site-packages/sklearn/base.py:311: UserWarning: Trying to unpickle estimator LinearSVC from version 0.21.2 when using version 0.19.1. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)
/Users/rodri/anaconda/anaconda3/lib/python3.6/site-packages/sklearn/base.py:311: UserWarning: Trying to unpickle estimator Pipeline from version 0.21.2 when using version 0.19.1. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)


Evaluation time 3.6658990383148193

Accuracy: 94.11% / 97.57% / 62.76% (total / known / unk)

g \ m	sp000	nc0s000	da0000	aq0000	fc	nc0p000	rg	np00000	fp	cc
sp000	14.30	0.00	-	0.03	-	-	0.00	-	-	-	
nc0s000	0.00	12.09	0.01	0.33	-	0.02	0.02	0.08	-	0.00	
da0000	-	0.08	9.53	-	-	0.00	-	0.00	-	-	
aq0000	0.00	0.34	-	6.61	-	0.15	0.01	0.05	-	-	
fc	-	-	-	-	5.85	-	-	-	-	-	
nc0p000	-	0.25	-	0.31	-	4.90	-	0.04	-	-	
rg	0.02	0.02	0.00	0.21	-	0.01	3.35	0.01	-	0.02	
np00000	0.00	0.21	-	0.08	-	0.01	-	3.26	-	0.00	
fp	-	-	-	-	-	-	-	-	3.55	-	
cc	0.00	-	-	0.01	-	-	0.05	0.00	-	3.34	


## Análisis de Errores

El script de evaluación eval.py calcula variables útiles para hacer análisis de errores:

- **error_count:** matriz de confusión completa.
- **error_sents:** matriz de índices de las oraciones en las que ocurre cada error.

Evaluamos el tagger con mejor performance (SVM) para hacer el análisis de errores.

Creamos un Pandas DF con las combinaciones de tags, el error count, el % y los indices de las oraciones donde ocurre ese error.

Imprimimos el las primeras filas

In [78]:
import pandas as pd

columns = []

for tag in tagsVocab:
    firstTag = tag
    for tag in tagsVocab:
        if tag == firstTag:
            continue            
        columns.append([firstTag, tag])
errors = pd.DataFrame(columns=['correct', 'predicted', 'n_errors', 'total', 'ind_error_sents'])
errors['correct'] = [col[0] for col in columns]
errors['predicted'] = [col[1] for col in columns]
n_errors = []
for tag in tagsVocab:
    firstTag = tag
    for tag in tagsVocab:
        if tag == firstTag:
            continue
        n_errors.append(error_count[firstTag][tag])  
errors['n_errors'] = n_errors
ind_error_sents = []
for tag in tagsVocab:
    firstTag = tag
    for tag in tagsVocab:
        if tag == firstTag:
            continue
        ind_error_sents.append(list(sorted(error_sents[firstTag][tag]))) 

errors['ind_error_sents'] = ind_error_sents
errors['total'] = round((errors['n_errors'] / total) * 100, 2)
errors.sort_values(by=['n_errors'], ascending = False, inplace=True)
errors.iloc[:10]

correct predicted  n_errors  total  \
241    aq0000   nc0s000       319   0.34   
82    nc0s000    aq0000       312   0.33   
723   nc0p000    aq0000       290   0.31   
721   nc0p000   nc0s000       241   0.25   
1291       cs  pr000000       225   0.24   
321   np00000   nc0s000       202   0.21   
1683       rg    aq0000       196   0.21   
248    aq0000   nc0p000       141   0.15   
3134  vmii000   vmip000       124   0.13   
2403  vmg0000    aq0000       112   0.12   

                                        ind_error_sents  
241   [32, 50, 70, 77, 112, 122, 127, 129, 135, 164,...  
82    [23, 26, 55, 60, 105, 106, 122, 132, 134, 137,...  
723   [141, 148, 151, 180, 231, 236, 241, 261, 266, ...  
721   [11, 44, 50, 82, 85, 113, 192, 252, 261, 265, ...  
1291  [2, 14, 32, 72, 106, 148, 151, 175, 185, 193, ...  
321   [6, 9, 12, 29, 36, 47, 49, 51, 55, 59, 61, 68,...  
1683  [31, 43, 135, 154, 177, 212, 221, 286, 326, 44...  
248   [42, 71, 103, 106, 126, 216, 223, 259, 342, 37...  
3134  [359, 415, 473, 525, 689, 797, 798, 850, 855, ...  
2403  [69, 111, 149, 226, 336, 372, 375, 420, 486, 5...

Recordamos los significados de los tag más frecuentes

| Tag    | Meaning  |
|------- | --------                    |
|sp000   |  Preposición
|nc0s000 |  Sustantivo común singular
|da0000  |  Artículo
|aq0000  |  Adjetivo Descriptivo
|fc      |  Coma
|np00000 |  Nombre Propio (sustantivo)
|nc0p000 |  Sustantivo común plural
|fp      |  Punto
|rg      |  Adverbio
|cc      |  Conjunción
|cs      |  Conjunción subordinante (que, porque, como)
|pr      |  Pronombre relativo (como, donde, cuando)


Vemos que los errores más frecuentes se dan cuando:

 - Un adjetivo se etiqueta como un sustantivo común singular
 - Un sustantivo común singular se etiqueta como un adjetivo
 - Un sustantivo común plural se etiqueta como un adjetivo
 - Un sustantivo común plural se etiqueta como singular
 - Una conjunción subordinante se etiqueta como  un pronombre relativo
 - Un adverbio se etiqueta como un adjetivo

## Caso más sencillo

El caso más sencillo de arreglar con features es el de **sustantivo singular y plural**.
Veamos una oración.

In [115]:
sent = sents[11]
words, tags = zip(*sent)  # separar las palabras de los tags
sent[:10]

[('Este', 'dd0000'),
 ('documento', 'nc0s000'),
 (',', 'fc'),
 ('que', 'pr000000'),
 ('ha', 'vaip000'),
 ('tomado', 'vmp0000'),
 ('como', 'sp000'),
 ('referencia', 'nc0s000'),
 ('la', 'da0000'),
 ('Comunidad_de_Madrid', 'np00000')]

In [113]:
df = pd.DataFrame(columns = ['real', 'predicted', 'word'])
df['real'], df['predicted'], df['word'] = tags,list(model.tag(words)), words
df['mistake'] = df['real'] == df['predicted']
df.loc[df['real'] == 'nc0p000']

real predicted          word  mistake
15  nc0p000   nc0p000      retornos     True
19  nc0p000   nc0p000   inversiones     True
40  nc0p000   nc0s000  innovaciones    False
44  nc0p000   nc0p000    beneficios     True

Vemos que taggeó mal la palabra 'innovaciones'. Chequeamos si es una palabra desconocida.

In [114]:
model.unknown('innovaciones')

True

Al ser una palabra desconocida solo se fija en las features que habíamos propuesto al principio: 

la palabra actual en minúsculas.
si la palabra actual empieza en mayúsculas.
si la palabra actual está en mayúsculas.
si la palabra actual es un número.
mismos features para la palabra anterior y para la siguiente.

Podemos agregar como feature si la palabra termina en 's' para aprovechar el caso de los plurales.

Modificamos un classfier.py donde agregamos como feature si la palabra termina con 's'. (guardamos la versión previa como V1)

Entrenamos nuevamente

In [1]:
%run tagging/scripts/train.py -c ancora-3.0.1es -m class -o trainedModels/v2SVM -t svm

training classifier with: svm  model
Extracting features

Executing pipeline
Training time 143.2172610759735
Saving...
All done!


In [3]:
%run tagging/scripts/eval.py -c ancora-3.0.1es -i trainedModels/v2SVM

Evaluation time 4.2035298347473145

Accuracy: 94.46% / 97.57% / 66.31% (total / known / unk)


Como vemos aumentó un poco la accuracy en palabras no vistas (62.76% vs 66.31%)

Chequeamos si la categoría que estábamos revisando mejoró (valor previo 0.25).


In [5]:
print((error_count['nc0p000']['nc0s000'] / total) * 100)

0.006331919204710948


Mejoró notablemente con la adición solamente de esa feature.